pennylane/labs/vibrational_ham/localize_modes.py

```diff
@@ -0,0 +1,206 @@
+            ugen[m1, m2] += params[idx]
+            ugen[m2, m1] -= params[idx]
+            idx += 1
+
+    return scipy.linalg.expm(ugen)
+
+
+def _params_cost(params, qmat, nmodes):
+    r"""Returns the cost function to be minimized for localized displacements."""
+    uparams = _params_to_unitary(params, nmodes)
+    qrot = _mat_transform(uparams, qmat)
+
+    return _pm_cost(qrot)
+
+
+def _q_normalizer(qmat):
+    r"""Returns the normalized displacement vectors."""
+    qnormalized = np.zeros_like(qmat)
+    nmodes = qmat.shape[2]
+
+    for m in range(nmodes):
+        m_norm = np.sum(np.abs(qmat[:, :, m]) ** 2)
+        qnormalized[:, :, m] = qmat[:, :, m] / np.sqrt(m_norm)
+
+    return qnormalized
+
+
+def _localization_unitary(qmat, rand_start=True):
+    r"""Returns the unitary matrix to localize the displacement vectors."""
+
+    nmodes = qmat.shape[2]
+    num_params = int(nmodes * (nmodes - 1) / 2)
+
+    if rand_start:
+        params = 2 * np.pi * np.random.rand(num_params)
+    else:
+        params = np.zeros(num_params)
+
+    qnormalized = _q_normalizer(qmat)
+
+    optimization_res = scipy.optimize.minimize(_params_cost, params, args=(qnormalized, nmodes))
+    if optimization_res.success is False:
+        warnings.warn("Mode localization finished unsuccessfully, returning normal modes...")
+        return _params_to_unitary(0 * params, nmodes), qmat
+
+    params_opt = optimization_res.x
+    uloc = _params_to_unitary(params_opt, nmodes)
+
+    qloc = _mat_transform(uloc, qmat)
+
+    return uloc, qloc
+
+
+def _localize_modes(freqs, disp_vecs, order=True):
+    r"""Performs the mode localization for a given set of frequencies and displacement vectors as
+    described in this `work <https://pubs.aip.org/aip/jcp/article-abstract/141/10/104105/74317/Efficient-anharmonic-vibrational-spectroscopy-for?redirectedFrom=fulltext>`_
+    """
+    nmodes = len(freqs)
+    hess_normal = np.zeros((nmodes, nmodes))
+    for m in range(nmodes):
+        hess_normal[m, m] = freqs[m] ** 2
+
+    natoms, _ = np.shape(disp_vecs[0])
+
+    qmat = np.zeros((natoms, 3, nmodes))
+    for m in range(nmodes):
+        dvec = disp_vecs[m]
+        for i in range(natoms):
+            for alpha in range(3):
+                qmat[i, alpha, m] = dvec[i, alpha]
+
+    uloc, qloc = _localization_unitary(qmat)
+    hess_loc = uloc.transpose() @ hess_normal @ uloc
+    loc_freqs = np.sqrt(np.array([hess_loc[m, m] for m in range(nmodes)]))
+
+    if order is True:
+        loc_perm = np.argsort(loc_freqs)
+        loc_freqs = loc_freqs[loc_perm]
+        qloc = qloc[:, :, loc_perm]
+        uloc = uloc[:, loc_perm]
+
+    return loc_freqs, qloc, uloc
+
+
+def localize_normal_modes(results, freq_separation=[2600]):
+    """
+    Arguments: results dictionary obtained from harmonic_analysis
+    separates frequencies at each point in freq_separation array
+
+    Returns: new dictionary with information for localized modes
+    """
```

Comment:
docstring is a little confusing. No worries though just a draft I figure it's still WIP.